vm_insnhelper.c

```diff
@@ -5382,7 +5382,11 @@ vm_opt_ltlt(VALUE recv, VALUE obj)
     }
     else if (RBASIC_CLASS(recv) == rb_cString &&
 	     BASIC_OP_UNREDEFINED_P(BOP_LTLT, STRING_REDEFINED_OP_FLAG)) {
-	return rb_str_concat(recv, obj);
+	if (LIKELY(RB_TYPE_P(obj, T_STRING))) {
+	    return rb_str_buf_append(recv, obj);
```

Comment:
Any insight why this is faster?
It seems only `RB_INTEGER_TYPE_P(str2)` and `StringValue(str2)` checks less, and I'd think those are fairly fast.

Reply:
If I remember correctly my profiling results (this patch is extracted from https://github.com/ruby/ruby/pull/6072, which I worked on last week), one of the main factor is the stupid amount of time we extract the encoding from the string.

So by jumping directly to `rb_str_buf_append`, we save the two call you mentioned, but also 2 function calls, and a fairly slow call to `get_encoding -> get_actual_encoding`.

Several of the other optimizations from #6072 that I need to cleanup also involve avoiding to go fetch the `rb_encoding *` and instead do some shortcut with the `encindex` for common encodings (utf-8, ascii, binary).

I think there's a bunch of low hanging fruits in `string.c` if we assume that 99.99% of strings are one of these 3 encodings.